Criar críticas para impedir Desativar ou Excluir Tipos de Hipóteses Legais utilizados na Administração do Módulo.

**Sua solicitação de melhoria está relacionada a um problema? Por favor descreva.**

O módulo não está utilizando eventos de bloqueio da API de módulos para impedir Desativações e Exclusões de Tipos de Hipóteses Legais, utilizadas nas diversas telas de Administração do Módulo PEN, o que acaba gerando erros no uso do módulo com mapeamentos feitos com registros desativados ou excluídos, onde na prática, acaba ocasionando erros que são difíceis de descobrir.

-----

**Descreva a solução que você gostaria**

O módulo PEN tem que implementar mensagens de crítica utilizando eventos de bloqueio da API de módulos para impedir Desativar ou Excluir Tipos de Hipóteses Legais e de Restrição Padrão utilizadas nas diversas telas de Administração do Módulo PEN.

Desativar:
- [x] Hipóteses Legais 

Excluir:
- [x] Hipóteses Legais

-----

Bom dia @NinetiesPaul ...realizei a aplicação do PR 359 no SEI-OFICIAL/MYSQL...No diretório SEI rodei o 

# make up e coreu tudo certo:

$ make up
docker-compose up -d
Starting sei-401215_solr_1      ... done
Starting sei-401215_jod_1       ... done
Starting sei-401215_memcached_1 ... done
Starting sei-401215_smtp_1      ... done
Starting sei-401215_database_1  ... done
sei-401215_httpd_1 is up-to-date
make check-sei-isalive


Aguardando inicialização do ambiente de desenvolvimento...
Tentativa 1/15
Tentativa 2/15
Página de login carregada!
make[1]: Saindo do diretório '/home/toor/Documentos/ministerio_economia/aplicacao_sei/sei_versoes/sei-4.0.12.15'

Mas, quando tento subir o módulo do TRAMITA (mod-sei-pen) ocorre o erro que segue:

$ make destroy; make up; make install
Stopping funcional_org2-database_1  ... done
Stopping funcional_org1-database_1  ... done
Stopping funcional_org2-memcached_1 ... done
Stopping funcional_org1-memcached_1 ... done
Removing funcional_org2-database_1  ... done
Removing funcional_php-test-unit_1  ... done
Removing funcional_smtp_1           ... done
Removing funcional_org1-database_1  ... done
Removing funcional_org2-memcached_1 ... done
Removing funcional_org1-memcached_1 ... done
Removing network funcional_default
docker-compose -f tests_super/funcional/docker-compose.yml --env-file tests_super/funcional/.env up -d
Creating network "funcional_default" with the default driver
Creating funcional_org2-memcached_1 ... 
Creating funcional_org1-memcached_1 ... 
Creating funcional_php-test-unit_1  ... 
Creating funcional_smtp_1           ... 
Creating funcional_org2-memcached_1 ... done
Creating funcional_org1-memcached_1 ... done
Creating funcional_php-test-unit_1  ... done

Creating funcional_org2-database_1  ... done
Creating funcional_org1-database_1  ... done

ERROR: for smtp  Cannot start service smtp: driver failed programming external connectivity on endpoint funcional_smtp_1 (1352456853ee1cd47d9d3678812ffcaef7c8ab2b92733ddd84d2daaa2b096e9a): Bind for 0.0.0.0:1080 failed: port is already allocated
ERROR: Encountered errors while bringing up the project.
make: *** [Makefile:164: up] Erro 1

Testando se a pagina de login responde para o org1...
Tentativa 1/15
Tentativa 2/15
Tentativa 3/15
Tentativa 4/15
Tentativa 5/15
Tentativa 6/15
Tentativa 7/15
Tentativa 8/15
Tentativa 9/15
Tentativa 10/15
Tentativa 11/15
Tentativa 12/15
Tentativa 13/15
Tentativa 14/15
Tentativa 15/15
Pagina de login do org1 nao encontrada. Verifique...
make: *** [Makefile:64: check-isalive] Erro 1

Alguma orientação?

--- a/src/PENIntegracao.php
+++ b/src/PENIntegracao.php
@@ -188,14 +188,24 @@ class PENIntegracao extends SeiIntegracao
   {
     $excecao = new InfraException();
     foreach ($arrObjHipoteseLegalAPI as $objHipoteseLegalAPI) {
-      $objPenHipoteseLegalDTO = new PenHipoteseLegalDTO();
+      $objPenHipoteseLegalDTO = new PenRelHipoteseLegalDTO();
       $objPenHipoteseLegalDTO->setNumIdHipoteseLegal($objHipoteseLegalAPI->getIdHipoteseLegal());
       $objPenHipoteseLegalDTO->retNumIdHipoteseLegal();
-      $objPenHipoteseLegalDTO->retStrNome();
 
-      $objPenHipoteseLegalRN = new PenHipoteseLegalRN();
-      $objPenHipoteseLegalDTO = $objPenHipoteseLegalRN->consultar($objPenHipoteseLegalDTO);
-      if (!is_null($objPenHipoteseLegalDTO)) {
+      $objPenRelHipoteseLegalEnvioRN = new PenRelHipoteseLegalEnvioRN();
+      $objPenRelHipoteseLegalEnvioDTO = $objPenRelHipoteseLegalEnvioRN->consultar($objPenHipoteseLegalDTO);
+
+      $objPenRelHipoteseLegalRecebidoRN = new PenRelHipoteseLegalRecebidoRN();
+      $objPenRelHipoteseLegalRecebidoDTO = $objPenRelHipoteseLegalRecebidoRN->consultar($objPenHipoteseLegalDTO);
+      
+      if (!is_null($objPenRelHipoteseLegalEnvioDTO) || !is_null($objPenRelHipoteseLegalRecebidoDTO)) {
+
+        $objPenHipoteseLegalDTO = new PenHipoteseLegalDTO();
+        $objPenHipoteseLegalDTO->setNumIdHipoteseLegal($objHipoteseLegalAPI->getIdHipoteseLegal());
+        $objPenHipoteseLegalDTO->retStrNome();
+
+        $objPenHipoteseLegalRN = new PenHipoteseLegalRN();
+        $objPenHipoteseLegalDTO = $objPenHipoteseLegalRN->consultar($objPenHipoteseLegalDTO);
         $excecao->lancarValidacao('Não é permitido excluir ou desativar a hipotese legal "' . $objPenHipoteseLegalDTO->getStrNome() . '"');
       }
     }

--- a/src/PENIntegracao.php
+++ b/src/PENIntegracao.php
@@ -191,6 +191,7 @@ class PENIntegracao extends SeiIntegracao
       $objPenHipoteseLegalDTO = new PenRelHipoteseLegalDTO();
       $objPenHipoteseLegalDTO->setNumIdHipoteseLegal($objHipoteseLegalAPI->getIdHipoteseLegal());
       $objPenHipoteseLegalDTO->retNumIdHipoteseLegal();
+      $objPenHipoteseLegalDTO->setNumMaxRegistrosRetorno(1);
 
       $objPenRelHipoteseLegalEnvioRN = new PenRelHipoteseLegalEnvioRN();
       $objPenRelHipoteseLegalEnvioDTO = $objPenRelHipoteseLegalEnvioRN->consultar($objPenHipoteseLegalDTO);

--- a/src/PENIntegracao.php
+++ b/src/PENIntegracao.php
@@ -174,6 +174,32 @@ class PENIntegracao extends SeiIntegracao
     return array($strAcoesProcedimento);
   }
 
+  public function excluirHipoteseLegal($arrObjHipoteseLegalDTO)
+  {
+    $this->validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalDTO);
+  }
+
+  public function desativarHipoteseLegal($arrObjHipoteseLegalDTO)
+  {
+    $this->validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalDTO);
+  }
+
+  public function validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalAPI)
+  {
+    $excecao = new InfraException();
+    foreach ($arrObjHipoteseLegalAPI as $objHipoteseLegalAPI) {
+      $objPenHipoteseLegalDTO = new PenHipoteseLegalDTO();
+      $objPenHipoteseLegalDTO->setNumIdHipoteseLegal($objHipoteseLegalAPI->getIdHipoteseLegal());
+      $objPenHipoteseLegalDTO->retNumIdHipoteseLegal();
+      $objPenHipoteseLegalDTO->retStrNome();
+
+      $objPenHipoteseLegalRN = new PenHipoteseLegalRN();
+      $objPenHipoteseLegalDTO = $objPenHipoteseLegalRN->consultar($objPenHipoteseLegalDTO);
+      if (!is_null($objPenHipoteseLegalDTO)) {
+        $excecao->lancarValidacao('Não é permitido excluir ou desativar a hipotese legal "' . $objPenHipoteseLegalDTO->getStrNome() . '"');
+      }
+    }
+  }
 
   public function montarIconeControleProcessos($arrObjProcedimentoAPI = array())
   {

--- a/src/PENIntegracao.php
+++ b/src/PENIntegracao.php
@@ -176,15 +176,15 @@ class PENIntegracao extends SeiIntegracao
 
   public function excluirHipoteseLegal($arrObjHipoteseLegalDTO)
   {
-    $this->validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalDTO);
+    $this->validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalDTO, 'exclusão');
   }
 
   public function desativarHipoteseLegal($arrObjHipoteseLegalDTO)
   {
-    $this->validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalDTO);
+    $this->validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalDTO, 'inativação');
   }
 
-  public function validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalAPI)
+  public function validarExcluirDesativarHipoteseLegal($arrObjHipoteseLegalAPI, $strAcao)
   {
     $excecao = new InfraException();
     foreach ($arrObjHipoteseLegalAPI as $objHipoteseLegalAPI) {
@@ -206,7 +206,11 @@ class PENIntegracao extends SeiIntegracao
 
         $objPenHipoteseLegalRN = new PenHipoteseLegalRN();
         $objPenHipoteseLegalDTO = $objPenHipoteseLegalRN->consultar($objPenHipoteseLegalDTO);
-        $excecao->lancarValidacao('Não é permitido excluir ou desativar a hipotese legal "' . $objPenHipoteseLegalDTO->getStrNome() . '"');
+        $nome = $objPenHipoteseLegalDTO->getStrNome();
+        $excecao->lancarValidacao($this->getNome().": 
+          A $strAcao da hipótese legal $nome não é permitida. 
+          A referida hipótese legal está relacionada a uma hipótese legal do Tramita."
+        );
       }
     }
   }

--- a/src/rn/PenHipoteseLegalRN.php
+++ b/src/rn/PenHipoteseLegalRN.php
@@ -53,4 +53,15 @@ class PenHipoteseLegalRN extends InfraRN
       throw new InfraException('Erro consultando hipotese legal.', $e);
     }
   }
+
+  protected function consultarConectado(PenHipoteseLegalDTO $objDTO)
+  {
+    try {
+      //Valida Permissao
+      $objBD = new GenericoBD(BancoSEI::getInstance());
+      return $objBD->consultar($objDTO);
+    } catch (Exception $e) {
+      throw new InfraException('Erro ao consultar Hipotese Legal.', $e);
+    }
+  }
 }

